### PR TITLE
SqlObject Factory rewrite

### PR DESCRIPTION
**Pull request by hgschmie**
- factor all the Generator specific code into GeneratorSqlObjectFactory, a new ExtensionFactory that deals specifically with the generated classes.
- strip all code that is not specific to the SqlObject Proxy extensions out of SqlObjectFactory
- register both factories in SqlObjectModule, so no user visible changes
- remove the specific code for generated classes (initData and lookupMethod) out of the SqlObjectInitData, this is now in the generator factory. Leave deprecated stubs in place for old, non-compiled classes. Those stubs will go away
- move the threadlocal that holds the init data into the generator factory
- move all init code for SqlObjectInitData into the class itself
- move duplicate superTypes code also into JdbiClassUtils

Even if the work stops here, this is a big improvement to make the SqlObject code actually understandable. The code is not that complicated, it was just split in very complicated ways and hard to follow. The PR splits it into three major pieces:

- everything that deals with the Proxy SqlObjects
- everything that deals with the generated SqlObjects
- everything that initializes the SqlObjectInitData object and its cache

When this is done, it should be possible to address the whole `invokeInContext` stuff on the HandleSupplier, the plan is to move this out of the handle supplier and into a wrapper around the method invocations itself, which would then allow things like annotations on default methods. Still a work in progress.

### CODE REVIEW

**Comment by hgschmie** [comment]
@stevenschlansker, as a major user of the generator, this would really benefit from some real world testing. I don't think this PR broke the generator and the generated code and all the tests pass, but the test suite is pretty light when it comes to the generator.

File: core/src/main/java/org/jdbi/v3/core/internal/JdbiClassUtils.java [hunk 1]
```diff
@@ -56,4 +67,11 @@ public static <T> T createInstance(Class<T> factoryClass) {
             throw new IllegalStateException(format("Unable to instantiate class %s", factoryClass), e);
         }
     }
+
+    public static Stream<Class<?>> superTypes(Class<?> type) {
+        Class<?>[] interfaces = type.getInterfaces();
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/core/internal/JdbiClassUtils.java:72]: Should this helper method emit duplicate super types? Like if A implements B,C and B implements C, you would emit C twice. (I realize this is not a new problem, but if we provide a helper, maybe we should specify it one way or the other.)

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java [hunk 2]
```diff
@@ -87,182 +64,29 @@ private boolean looksLikeSqlObject(Class<?> extensionType) {
...
+        final SqlObjectInitData sqlObjectInitData = SqlObjectInitData.lookup(extensionType, handleSupplier.getConfig());
+        final ConfigRegistry instanceConfig = sqlObjectInitData.configureInstance(handleSupplier.getConfig().createCopy());
+
         instanceConfig.get(Extensions.class).onCreateProxy();
 
         Map<Method, Supplier<InContextInvoker>> handlers = new HashMap<>();
         final Object proxy = Proxy.newProxyInstance(
                 extensionType.getClassLoader(),
-                new Class[] {extensionType},
+                new Class[]{extensionType},
```
Comment (stevenschlansker) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java:85]: the style I use would usually keep the space between `[] {`

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/internal/SqlObjectInitData.java [hunk 3]
```diff
@@ -121,6 +102,139 @@ public Supplier<InContextInvoker> lazyInvoker(Object target, Method method, Hand
...
+
+        // tostring can not be constant, different for every object
+        Handler toStringHandler = (target, args, handleSupplier) ->
+                "Jdbi sqlobject proxy for " + sqlObjectType.getName() + "@" + Integer.toHexString(target.hashCode());
+
+        addMethodHandler(handlerMap, toStringHandler, Object.class, "toString");
+        addMethodHandler(handlerMap, EQUALS_HANDLER, Object.class, "equals", Object.class);
+        addMethodHandler(handlerMap, HASHCODE_HANDLER, Object.class, "hashCode");
+
+        // this should probably have a check whether the sqlObjectType actually implements SqlObject
```
Comment (stevenschlansker) [inline sqlobject/src/main/java/org/jdbi/v3/sqlobject/internal/SqlObjectInitData.java:167]: I don't think it is harmful to have extra handler methods here that you can't actually call.

**Review by stevenschlansker — APPROVED** [review]

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/internal/SqlObjectInitData.java (hunk 3 above)
Comment (hgschmie) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/internal/SqlObjectInitData.java:167]: that comment has always been there.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java (hunk 2 above)
Comment (hgschmie) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java:85]: weird.

File: core/src/main/java/org/jdbi/v3/core/internal/JdbiClassUtils.java (hunk 1 above)
Comment (hgschmie) [reply core/src/main/java/org/jdbi/v3/core/internal/JdbiClassUtils.java:72]: I added a deduplication step.

File: sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java (hunk 2 above)
Comment (hgschmie) [reply sqlobject/src/main/java/org/jdbi/v3/sqlobject/SqlObjectFactory.java:85]: The answer seems to be "IntelliJ has a new checkbox and it was off" :-P

**Comment by stevenschlansker** [comment]
The test suite is light, but does have pretty good coverage of a lot of different cases. Will let you know if I see any problems.
